### DISCUSSION TITLE
Micro frontend next.js CORS issue

### DISCUSSION THREAD

**Discussion by StephenHolmes14** [vercel/next.js#16942] [Help]
Hello,

I'm trying to implement next.js as part of a micro front-end. I'm having issues getting around CORS for the static files next js hosts.

I've got this inside my `next.config.js` which seems to work for the pages but not the static files..

```
module.exports = {
  basePath: '/base',
  async headers() {
    return [
      {
        source: '/:path*',
        headers: [
          {
            key: 'Access-Control-Allow-Origin',
            value: '*',
          },
        ],
      }
    ]
  },
}
```

Example file not working:
```
Access to fetch at 'http://localhost:3000/base/_next/static/css/3909390f96ab4d9bf9ee.css' from origin 'http://localhost:3001' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
VM272:1 GET http://localhost:3000/base/_next/static/css/3909390f96ab4d9bf9ee.css net::ERR_FAILED
```

If anyone has done this or knows if it's even possible please help 🙏

**Comment by gautamkr1876** [comment]
are you using ModuleFederationPlugin to implement microfrontend ?

**Reply by jeevanandham5** [reply]
i am getting the same issue   and yes i am using  ModuleFederationPlugin

**Comment by nguyenthanhan201** [comment]
<img width="1474" alt="Screenshot 2024-01-16 at 13 18 49" src="https://github.com/vercel/next.js/assets/80932055/1e9622f5-3b52-4647-b07c-0f7bbb237773">
i have same error, when i use react(vite) like app-shell and nextjs is a remote-module. How i can fix that error
Note: Nextjs (@module-federation/nextjs-mf), Vite (@originjs/vite-plugin-federation)

**Reply by ManuelcsNuant** [reply]
@nguyenthanhan201 I'm getting exactly this issue, react vite is the app shell and next14 as a remote-module and getting same issue

**Comment by m-avagyan** [comment]
I configured my next.js middleware to handle accepted origins and return website or route only if origin type is accepted for production mode. For local development I set origin value as * because no need to handle CORS for localhost. You can copy paste this code to your middleware.ts file.

```typescript
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

const isProduction = process.env.DATASET === 'production';
const corsOrigins = isProduction
    ? [
          'domain1',
          'domain2',
          'domain3',
          'domain4',
      ]
    : ['*'];

export function middleware(req: NextRequest) {
    const origin = req.headers.get('origin');
    const res = NextResponse.next();

    // Set CORS headers
    res.headers.set('Access-Control-Allow-Credentials', 'true');
    res.headers.set('Access-Control-Allow-Methods', 'GET,DELETE,PATCH,POST,PUT');
    res.headers.set(
        'Access-Control-Allow-Headers',
        'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version'
    );

    if (corsOrigins.includes('*') || corsOrigins.includes(origin || '')) {
        res.headers.set('Access-Control-Allow-Origin', origin || '*');
    }

    return res;
}

export const config = {
    matcher: '/api/:path*', // Match all API routes
};
```

Please modify the `corsOrigins` array value (add all your needed domains).

**Reply by tejas-1206** [reply]
Hi @m-avagyan , this is an interesting approach. May I know what your MFE project setup was ?
My host app is a Vite-based ReactJS app and uses `originjs/vite-plugin-federation` while my remote is a NextJS app uses `module-federation/nextjs-mf` to expose the required files.

The middleware you shared made sense and I added it to my `middleware.ts` file, but it did not resolve the issue.
I still faced the same issue on my host. I did try by serving my remote locally as well.
Error :
```bash
Access to script at 'http://localhost:3000/_next/static/chunks/remoteEntry.js' from origin 'https://test.internalDomain.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

Not able to understand why though. Could you please point me in the right direction where I can find a solution to this?
I made sure that `http://localhost:3000/_next/static/chunks/remoteEntry.js` is returning the chuck required for consumption.

-------------------------------------------------------------------------------------------------------------------
One interesting thing that I found was, in my host's fed mod config, if I set `format: "var"`, the CORS issue gets resolved, but then it throws a new set of errors saying 
```bash
Uncaught TypeError: Cannot read properties of null (reading 'useContext')
Uncaught TypeError: Cannot read properties of null (reading 'useState')
```
Config in my host for your ref:
```ts
federation({
        name: "teams",
        remoteType: "module",
        remotes: {
          auth: {
            externalType: "url",
            // external: `${env.VITE_AUTH_REMOTE_ENTRY_BASE_URL}/_next/static/chunks/remoteEntry.js`,
            external:
              "http://localhost:3000/_next/static/chunks/remoteEntry.js",
            format: "var",
            from: "webpack"
          }
        },
        shared: ["react", "react-dom", "axios"]
      })
```

Config in my remote for your ref:
```mjs
webpack(config) {
    config.plugins.push(
      new NextFederationPlugin({
        name: "auth",
        library: {
          type: "module",
        },
        remotes: {},
        filename: "static/chunks/remoteEntry.js",
        exposes: {
          "./Auth": "./components/Auth/index.tsx",
          "./logout": "./lib/logout/index.ts",
        },
        extraOptions: {
          debug: false, // `false` by default
          exposePages: false, // `false` by default
        },
      })
    );

    return config;
  },
  ```
I wonder why.

**Comment by ArunPSreenivasan** [comment]
create a server.js file to manage cors. this helped me to fix the same issue

```
const express = require("express");
const next = require("next");
const cors = require("cors");

const dev = process.env.NODE_ENV !== "production";
const app = next({ dev });
const handle = app.getRequestHandler();

const port = process.env.PORT || 3001; // Specify your port number here

app.prepare().then(() => {
  const server = express();

  // Enable CORS for all routes
  server.use(cors());

  // Handle all requests with Next.js
  server.all("*", (req, res) => {
    return handle(req, res);
  });

  // Start the server on the specified port
  server.listen(port, (err) => {
    if (err) throw err;
    console.log(`> Ready on http://localhost:${port}`);
  });
});
```

**Reply by bdanielmr** [reply]
> create a server.js file to manage cors. this helped me to fix the same issue
> 
> ```
> const express = require("express");
> const next = require("next");
> const cors = require("cors");
> 
> const dev = process.env.NODE_ENV !== "production";
> const app = next({ dev });
> const handle = app.getRequestHandler();
> 
> const port = process.env.PORT || 3001; // Specify your port number here
> 
> app.prepare().then(() => {
>   const server = express();
> 
>   // Enable CORS for all routes
>   server.use(cors());
> 
>   // Handle all requests with Next.js
>   server.all("*", (req, res) => {
>     return handle(req, res);
>   });
> 
>   // Start the server on the specified port
>   server.listen(port, (err) => {
>     if (err) throw err;
>     console.log(`> Ready on http://localhost:${port}`);
>   });
> });
> ```

hi this server how setup on my remote app or my host app
